design/handle-backup-of-volumes-by-resources-filters.md

```diff
@@ -0,0 +1,211 @@
+ 
+## Background
+As of Today, Velero has lots of filters to handle (backup or skip backup) resources including resources filters like `IncludedNamespaces, ExcludedNamespaces`, label selectors like `LabelSelector, OrLabelSelectors`, annotation like `backup.velero.io/must-include-additional-items` etc. But it's not enough flexible to handle volumes, we need one generic way to filter volumes.
+ 
+## Goals
+- Introducing one flexible way to filter volumes.
+ 
+## Non Goals
+- We only handle volumes for backup and do not support restore.
+- Currently, only handle volumes, does not support other resources.
+- Only environment-unrelated and platform-independent general volumes attributes are supported, do not support volumes attributes related to a specific environment.
+ 
+## Use-cases/Scenarios
+### A. Skip backup volumes by some attributes
+Users want to skip PV with the requirements:
+- option to skip all PV data
+- option to skip specified PV type (RBD, NFS)
+- option to skip specified PV size
+- option to skip specified storage-class
+- option to skip folders
+ 
+### B. Accurately select target volume to backup
+Some volumes are only used for logging while others volumes are for DBs, and only need to backup DBs data. users need `one specific resource seletor` to accurately select target volumes to backup.
+ 
+## High-Level Design
+Add a new flag `from-file` when executing `velero backup create`, which imports the defined resources filters in one JSON file. the JSON file including all defined filter rules for the current backup.
+ 
+When Velero handles volumes backup should respect the filter rules defined in the imported JSON file.
+ 
+## Detailed Design
+The resources filters rules should contain both `include` and `exclude` rules.
+ 
+For the rules on `one specific resource selector`, we introduced a `GVRN` way of resources filters, for resources are identified by their resource type and resource name, or GVRN.
+
+Here we call it `GVRN Selector` which exactly matches the resources to be handled.
+
+For the attributes on `some specific volumes attributes`, we basically follow the defined data struct [PersistentVolumeSpec](https://github.com/kubernetes/kubernetes/blob/v1.26.0/pkg/apis/core/types.go#L304), and only handle partial common fields of it currently.
+ 
+Here we call it `Volumes Attributes Selector`, which matches volumes with the same attributes defined.
+ 
+### filter fields format
+The filter JSON config file would look like this:
+```
+{
+   "resources": {
+       "include": {
+           "groupResource": "/persistentvolumes",
+           "namespacedNames": "/nginx-logs"
+       }
+       "exclude": {
+           "groupResource": "apps/deployments",
+           "namespacedNames": "velero/velero"
+       }
+   },
+   "storage": {
```

Comment:
It's better to combine the "pv" and "volume" in storage filters. Reasons:

- Users will need to consider the backup types (volume snapshot or pod volume backup?) when they use these filters, this will add unnecessary burden to users
- We will add more backup types that processing PVs in future, if we keep the backup type differences here, we will have to define new fields for the new backup types

Reply:
yes, I agree with you.